src/analysisd/analysisd.c

```diff
@@ -2394,3 +2546,156 @@ void w_init_queues(){
+        }
+
+        if (limits.eps) {
+
+            if (!old_timeframe && limits.timeframe){
+                /* First time */
+                os_calloc(limits.timeframe, sizeof(unsigned int), limits.circ_buf);
+
+            } else if (old_timeframe != limits.timeframe) {
+                /* Resize buffer */
+                unsigned int buf[old_timeframe];
+                memset(buf, 0, sizeof(unsigned int) * old_timeframe);
+                memcpy(buf, limits.circ_buf,  sizeof(unsigned int) * old_timeframe);
+                os_realloc(limits.circ_buf, limits.timeframe * sizeof(unsigned int), limits.circ_buf);
+
+                if (old_timeframe > limits.timeframe) {
+                    /* Shrink buffer */
+                    unsigned int offset = limits.current_cell + 1 > limits.timeframe ? limits.current_cell + 1 - limits.timeframe : 0;
+                    memcpy(limits.circ_buf, buf + offset, sizeof(unsigned int) * limits.timeframe);
+                    limits.current_cell = limits.current_cell + 1 >= limits.timeframe ? limits.timeframe - 1 : limits.current_cell;
+
+                } else {
+                    /* Expand buffer */
+                    memset(limits.circ_buf, 0, sizeof(unsigned int) * limits.timeframe);
+                    memcpy(limits.circ_buf, buf, sizeof(unsigned int) * old_timeframe);
+                }
+            }
+
+            limits.total_eps_buffer = 0;
+            for (unsigned int i = 0; i < limits.current_cell; i++) {
+                 limits.total_eps_buffer += limits.circ_buf[i];
+            }
+
+            limits.max_eps = limits.eps * limits.timeframe;
+
+            int current_value = 0;
+            sem_getvalue(&credits_eps_semaphore, &current_value);
+            unsigned int c_value = (unsigned int) current_value;
+
+            if (limits.max_eps > limits.total_eps_buffer) {
+                if (c_value < limits.max_eps - limits.total_eps_buffer) {
+                    generate_eps_credits(limits.max_eps - limits.total_eps_buffer - c_value);
+                } else if (c_value > limits.max_eps - limits.total_eps_buffer){
+                    clean_eps_credits(c_value + limits.total_eps_buffer - limits.max_eps);
+                }
+            } else if (current_value) {
+                clean_eps_credits(current_value);
+            }
+
+            limits.enabled = true;
+
+            minfo("eps limit enable, eps: %d, timeframe: %d, events per timeframe: %d", limits.eps, limits.timeframe, limits.max_eps);
+        }
+        else {
+            minfo("eps limit disabled");
+            limits_free();
+        }
+
+        w_mutex_unlock(&limit_eps_mutex);
+        cJSON_Delete(analysisd_limits);
+
+    } else if (err == LIMITS_FILE_NOT_FOUND && limits.enabled) {
+        minfo("eps limit disabled");
+        limits_free();
+    }
+}
+
+static void limits_free(void) {
+    if (limits.circ_buf) {
+        os_free(limits.circ_buf);
+    }
+    while(limits.wait_counter--) {
```

Comment:
```suggestion
    while (limits.wait_counter--) {
```
Also, shouldn't we protect this with the `wait_sem` lock?

Reply:
Done